kmod/muser.c

```diff
@@ -654,7 +654,7 @@ get_dma_map(struct muser_dev *mudev, struct vfio_dma_mapping *dma_map,
 	struct page **pages;
 
 	length = map->size;
-	pages = kmalloc_array(NR_PAGES(length), sizeof(*pages), GFP_KERNEL);
+	pages = vmalloc(NR_PAGES(length) * sizeof(*pages));
```

Comment:
I think using kvmalloc() or kvmalloc_array should be better, it first tries to allocate physically contiguous pages otherwise returns virtually contiguous pages.

Reply:
Why do we care if pages contiguous?

Reply:
kmalloc() is faster than vmalloc() and also the accesses to the kmalloc'ed memory is faster as additional translation is not required. So mostly in the non-huge page cases we can leverage the kmalloc'ed memory.
I do not have a strong opinion on this, its just a suggestion.

Reply:
Fair enough, I'll create an issue ti improve this in the future.